examples/Example17/src/ParticleGun.cc

```diff
@@ -0,0 +1,48 @@
+// SPDX-FileCopyrightText: 2023 CERN
+// SPDX-License-Identifier: Apache-2.0
+//
+#include "ParticleGun.hh"
+#include "G4ParticleGun.hh"
+#include "G4PhysicalConstants.hh"
+#include "G4PrimaryParticle.hh"
+#include "G4Event.hh"
+#include "Randomize.hh"
+#include "G4SystemOfUnits.hh"
+
+void ParticleGun::GenerateRandomPrimaryVertex(G4Event *aEvent, G4double aMinPhi, G4double aMaxPhi, G4double aMinTheta,
+                                              G4double aMaxTheta, std::vector<G4ParticleDefinition *> *aParticleList)
+{
+  // Choose a particle from the list
+  G4ParticleDefinition *aParticleDefinition = (*aParticleList)[(int)(G4UniformRand() * aParticleList->size())];
+  SetParticleDefinition(aParticleDefinition);
+
+  // Create a new vertex
+  //
+  auto *vertex = new G4PrimaryVertex(particle_position, particle_time);
+
+  // Create new primaries and set them to the vertex
+  //
+  G4double mass = particle_definition->GetPDGMass();
+  for (G4int i = 0; i < NumberOfParticlesToBeGenerated; ++i) {
+    auto *particle = new G4PrimaryParticle(particle_definition);
+
+    G4double phi                = (aMaxPhi - aMinPhi) * G4UniformRand() + aMinPhi;
+    G4double theta              = (aMaxTheta - aMinTheta) * G4UniformRand() + aMinTheta;
```

Comment:
Here you generate uniform in theta, but for a isotropic distribution (same probability per solid angle) you need to generate uniform in cos(theta):
```
  auto theta = std::acos(std::cos(aMinTheta) + G4UniformRand() * (std::cos(aMaxTheta) - std::cos(aMinTheta)));
```